Src/ILGPU/Runtime/ExchangeBuffer.tt

```diff
@@ -0,0 +1,480 @@
+        /// </summary>
+        /// <param name="buffer">The source buffer.</param>
+        public static implicit operator MemoryBuffer<T, TIndex>(
+            ExchangeBufferBase<T, TIndex> buffer)
+        {
+            Debug.Assert(buffer != null, "Invalid buffer");
+            return buffer.Buffer;
+        }
+
+        #endregion
+
+        #region IDisposable
+
+        /// <summary cref="DisposeBase.Dispose(bool)"/>
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                Buffer.Dispose();
+                cpuMemory.Dispose();
+            }
+            base.Dispose(disposing);
+        }
+
+        #endregion
+    }
+    <#for(int i = 1; i <= top; i++)
+    {#>
+    <# var indexName = $"Index{i}"; #>
+    <# var dimensionName = $"{i}D"; #>
+    <# var typeName = i == 1 ? "ExchangeBuffer" : $"ExchangeBuffer{dimensionName}"; #>
+    <# var arrayViewType = $"ArrayView<T, {indexName}>"; #>
+    <# var highLevelView =  i == 1 ? "ArrayView" : $"ArrayView{dimensionName}"; #>
+    <# var memoryBufferType = $"MemoryBuffer<T, {indexName}>"; #>
+    <# var highLevelMemBuf =  i == 1 ? "MemoryBuffer" : $"MemoryBuffer{dimensionName}"; #>
+    /// <summary>
+    /// <#= dimensionName #> implementation of <see cref="ExchangeBufferBase{T, TIndex}"/>
+    /// </summary>
+    /// <typeparam name="T">The element type.</typeparam>
+    /// <remarks>Members of this class are not thread safe.</remarks>
+    public sealed unsafe class <#= typeName #><T> : ExchangeBufferBase<T, <#= indexName #>>
+        where T : unmanaged
+    {
+        #region Instance
+
+        /// <summary>
+        /// Initializes this memory buffer.
+        /// </summary>
+        /// <param name="buffer">The underlying memory buffer.</param>
+        /// <param name="mode">The current buffer allocation mode.</param>
+        internal <#= typeName #>(<#= memoryBufferType #> buffer,
+            ExchangeBufferMode mode)
+            : base(buffer, mode)
+        {
+            <#if(i > 1)
+            {#>var baseView = new ArrayView<T>(CPUMemory, 0, buffer.Length);
+            CPUView = new <#= arrayViewType #>(baseView, buffer.Extent);
+            <#}
+            else
+            {#>CPUView = new ArrayView<T>(CPUMemory, 0, buffer.Length);
+            <#}#>
+
+            // Cache local data
+            Buffer = buffer;
+            NativePtr = buffer.NativePtr;
+            View = buffer.View;
+            Extent = buffer.Extent;
+        }
+
+        #endregion
+
+        #region Methods
+
+        /// <summary>
+        /// Gets this buffer as a higher level memory buffer.
+        /// </summary>
+        /// <returns> A <see cref="<#= highLevelMemBuf #>{T}"/> containing the data in this
+        /// exchanage buffer.</returns>
+        public <#= highLevelMemBuf #><T> As<#= highLevelMemBuf #>() => new <#= highLevelMemBuf #><T>(Buffer);
+
+        <#if(i > 1)
```

Comment:
@Yey007 For T4 templates, any extra whitespace is written to the output file.

You should move the opening `<#` to the start of the line. Then, either move the `if` statement to the start, or add whitespace before the `if` statement.

Note that Whitespace within `<#` and `#>` is fine, as this is not written to the output file.


```suggestion
<#if(i > 1)
```

```suggestion
<#        if(i > 1)
```

Reply:
Oh alright, I got it I think. Thanks.